apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -292,13 +303,133 @@ public void beforeEnd(long epochMicros) {
 
     @Override
     protected void afterEnd() {
-        this.tracer.endSpan(this);
+        if (transaction != null && transaction.isSpanCompressionEnabled() && parent != null) {
+            Span buffered = parent.bufferedSpan.get();
+            if (!isCompressionEligible()) {
+                if (buffered != null) {
+                    if (parent.bufferedSpan.compareAndSet(buffered, null)) {
+                        this.tracer.endSpan(buffered);
+                    }
+                }
+                this.tracer.endSpan(this);
+                return;
+            }
+            if (buffered == null) {
+                if (!parent.bufferedSpan.compareAndSet(null, this)) {
+                    // the failed update would ideally lead to a compression attempt with the new buffer,
+                    // but we're dropping the compression attempt to keep things simple
+                    this.tracer.endSpan(this);
+                }
+                return;
+            }
+            if (!buffered.tryToCompress(this)) {
+                if (!parent.bufferedSpan.compareAndSet(buffered, this)) {
+                    // the failed update would ideally lead to a compression attempt with the new buffer,
+                    // but we're dropping the compression attempt to keep things simple
+                    this.tracer.endSpan(buffered);
```

Comment:
the buffer is still set to `buffered` here, but it's being reported, I'm think that leads to multiple reports of the same span?

Reply:
I think you are (partly) right here. If we fail to update the buffer, the buffer is already set to something different than `buffered`. So I don't think `buffered` will be reported multiple times (in the future). But as the buffer is set to something different than `buffered`, `buffered` should already have been reported. So  I think `buffered` will be reported for the second time here.

@jackshirazi Do you agree/disagree?